drivers: adc_shell: rework commands

jakub-uC opened the pull request:
ADC shell commands can be only executed on the property label exiting in the device tree. It is realized by the dynamic subcommand returning the existing ADC Label.

Fixes #27350

Signed-off-by: Jakub Rzeszutko <jakub.rzeszutko@nordisemi.no>

jakub-uC commented:
@avigreen1978 may you please test it?

erwango commented:
^^ @avigreen1978

erwango commented on drivers/adc/adc_shell.c:
```diff
@@ -32,17 +32,41 @@
 #define ADC_HDL_LIST_ENTRY(inst)					\
 	{								\
-		.device_name = DT_INST_LABEL(inst),			\
+		.device_name = inst,					\
```
If I reading it right (otherwise, please discard this comment):
While this is working and doing the expected thing, the use of `inst` is misleading here.
In DT macros, we're usually calling INST the ordinal number of the node (2nd, 3rd, ..).
In `ADC_HDL_LIST_ENTRY`, the argument is actually the label property value ("ADC_X" string).

So, to clarify this piece of code, I'd witch argument to something less misleading: `node_label` or similar.

jakub-uC replied:
Sure, I will update the naming. It is a property label like `ADC_0` or `ADC_1`.

erwango commented:
@jakub-uC basic test ok

erwango reviewed (approved):
Approved from my side, but need ADC friends reviews.

nordic-krch commented on drivers/adc/adc_shell.c:
```diff
@@ -233,256 +214,198 @@ static int cmd_adc_acq(const struct shell *shell, size_t argc, char **argv)
-	SHELL_CMD(EXT0, NULL, "EXT0", cmd_adc_ref),
-	SHELL_CMD(EXT1, NULL, "EXT1", cmd_adc_ref),
-	SHELL_SUBCMD_SET_END /* Array terminated. */
+	SHELL_CMD_ARG(VDD_1, NULL, "VDD", cmd_adc_ref, 1, 0),
```
we could think at some point about shell extension which would allow creating subcommand with a dictionary. In that case argv would contain direct value. Something like
```
SHELL_STATIC_SUBCMD_DICT_CREATE(sub, dictionary_array)
```
where array contains string + value. String for tab handle, value passed to command handler. That would make code much compact in cases like this and there are many cases like this one.

jakub-uC replied:
May you please create an issue for that?

anangl commented on drivers/adc/adc_shell.c:
```diff
@@ -52,178 +72,139 @@ struct adc_hdl {
-			&adc_list[chosen_adc].channel_config);
+
+	if (IS_ENABLED(CONFIG_ADC_CONFIGURABLE_INPUTS)) {
+		adc->channel_config.input_positive = 1;
```
This should not be hard-coded. Users should be able to select the input to use. And this value may be even invalid for some drivers.

jakub-uC replied:
@anangl Do you think it is a good idea to add one more command for setting this value?

jakub-uC commented on drivers/adc/adc_shell.c:
```diff
@@ -32,17 +34,35 @@
+	"Configure resolution\n"		\
+	"Usage: resolution <resolution>\n"
+
+#define CMD_HELP_REF	"Configure resolution\n"
```
fixed

carlescufi commented:
@anangl please take another look

anangl commented on drivers/adc/adc_shell.c:
```diff
@@ -32,17 +34,45 @@
+
+#define CMD_HELP_CH_ID				\
+	"Configure channel id\n"		\
+	"Usage: channel <channel_id>\n"
```
I guess this should be
```suggestion
	"Usage: id <channel_id>\n"
```

jakub-uC replied:
fixed

anangl commented on drivers/adc/adc_shell.c:
```diff
@@ -52,178 +82,201 @@ struct adc_hdl {
+	{ .string = "ADC_GAIN_8", .gain = ADC_GAIN_8 },
+	{ .string = "ADC_GAIN_16", .gain = ADC_GAIN_16 },
+	{ .string = "ADC_GAIN_32", .gain = ADC_GAIN_32 },
+	{ .string = "ADC_GAIN_64", .gain = ADC_GAIN_64 }
```
Could you add `ADC_GAIN_128` as well (here and in `sub_gain_cmds`), as it was added in `adc.h`?

jakub-uC replied:
done

pabigot commented:
Last dev review we'd discussed #28566 and identified a generic way to associate devices with specific APIs.  @jakub-uC are you pursuing that approach?  Would doing that supersede this proposal?

It's a little hard to understand what this is doing since it packs a bunch of refactoring/reformatting along with the described device access change into a single commit.  I think there's a device access change in there somewhere.

jakub-uC commented:
@pabigot : I think we can merge this PR and next it could be updated according to your proposal.

The most important parts of the code of this change are here:

```
static void cmd_adc_dev_get(size_t idx, struct shell_static_entry *entry)
{
	/* -1 because the last element in the list is a "list terminator" */
	if (idx < ARRAY_SIZE(adc_list) - 1) {
		entry->syntax  = adc_list[idx].device_label;  // <=== HERE
		entry->handler = NULL;
		entry->subcmd  = &sub_adc_cmds;
		entry->help    = "Select subcommand for ADC property label.\n";
	} else {
		entry->syntax  = NULL;
	}
}
```

Here: `entry->syntax  = adc_list[idx].device_label;` I need an array of strings with active ADCs. This string is used to: prompt a device name in the shell and bind an ADC device in the command handler.

A list of strings is created here:

```
#define INIT_MACRO() DT_INST_FOREACH_STATUS_OKAY(NODE_LABELS) "NA"

/* This table size is = ADC devices count + 1 (NA). */
static struct adc_hdl {
	char *device_label;
	struct adc_channel_cfg channel_config;
	uint8_t resolution;
} adc_list[] = {
	FOR_EACH(ADC_HDL_LIST_ENTRY, (,), INIT_MACRO())
};
```
I believe it is also part of the code that you would need to change according to your proposal.

pabigot reviewed (commented):
From https://github.com/zephyrproject-rtos/zephyr/pull/28599#issuecomment-706090444

> @pabigot : I think we can merge this PR and next it could be updated according to your proposal.
> 
> The most important parts of the code of this change are here:

This ideally would have been done in multiple, focused commits. Among all the re-ordering of help and structure definitions, conversion from array index to pointer, etc it's not at all clear what was done and why.  Since this already has three approvals I'm not going to block this, but I'm not going to approve it because it's not worth the time to try to figure it all out.

nordic-krch commented:
@jakub-uC what about using dictionary commands? Are you plan it in this PR?

jakub-uC commented:
@nordic-krch : Yes I am reworking that and splitting to different commits. 
When this PR was submitted there were no dictionary commands on the master branch.

Commit:
adc_shell: introduce dictionary commands

Changed gain and reference commands to dictionary commands.
This change removes an obsolete look-up table (string <-> value)
for gain and reference commands.
Now, each modification of gain or reference value will require only
a dictionary command update.

Signed-off-by: Jakub Rzeszutko <jakub.rzeszutko@nordicsemi.no>

Commit:
adc_shell: use shell to validate arguments count

Use dedicated shell macros so argument count can be validated before
the command handler is executed. This change simplifies the command
handlers implementation inside the adc_shell file.

Signed-off-by: Jakub Rzeszutko <jakub.rzeszutko@nordicsemi.no>

Commit:
adc_shell: extend channel command

The channel command has been extended with subcommands: id, positive,
and negetive. Some boards require positive input configuration before
measurement can be started.

Signed-off-by: Jakub Rzeszutko <jakub.rzeszutko@nordicsemi.no>

Commit:
adc_shell: rework commands

ADC shell commands can be only executed on the property label
existing in the device tree. It is realized by the dynamic subcommand
returning existing ADC Label.

Signed-off-by: Jakub Rzeszutko <jakub.rzeszutko@nordicsemi.no>

jakub-uC commented:
@pabigot I've tried to make reasonable commits. The most important is the last commit: `adc_shell: rework commands`

In order to prompt valid ADC labels I've implemented dynamic command that is realized by the function `static void cmd_adc_dev_get(size_t idx, struct shell_static_entry *entry)`. What this function needs for proper work is an array of valid ADC labels that can be used to bind the ADC device.

If we know that the ADC label, for example, `ADC_0`, is prompted by the shell as a subcommand it is not needed in the commands handler to check if the device is on the list. This is why commands were simplified.
